[develop] Replace shell-based WE2E scripts with python versions

mkavulich opened the pull request:
## DESCRIPTION OF CHANGES: 
This PR improves on the new `./run_WE2E_tests.py` script (introduced in #558), implementing all the features present in the previous shell-based workflow. Some new files are also introduced for better organization and additional functionality:

 - `tests/WE2E/utils.py` This is a collection of functions used by other scripts, contained here to avoid circular dependencies.
 - `tests/WE2E/WE2E_summary.py` Given an experiment directory or .yaml file, outputs a summary to screen of each experiment, its status, and the number of core hours used. It also prints a summary file with detailed information about each task for each experiment.
 - `tests/WE2E/print_test_info.py` Will print a file `WE2E_test_info.txt`, very similar to the legacy `WE2E_test_info.csv` with just a few minor format differences.

Any scripts can be run with the `-h` argument to print information about all available options (not including `utils.py`, which is not designed to be run stand-alone).

**With this PR, the old superseded shell-based tools are removed.** This will require action by anyone whose workflow relies on those tools. I will reach out to those that I know have automated testing based on these tools, but please let me know if any other capabilities or modifications are needed to keep existing workflows working.

### New options
 - `--procs=##` For monitoring and submitting jobs, this will call `rocotorun` and read rocoto database files in parallel for the specified number of parallel tasks. This can greatly speed up the submission of large test suites, such as the comprehensive tests. **Only use this option if you are sure you have access to the specified number of parallel cores**; python will behave badly if you try to over-subscribe your available tasks.
 - `--opsroot=` If test is for NCO mode, sets OPSROOT
 - `--print_test_info` Create a `WE2E_test_info.txt` file summarizing each test prior to starting experiment (False by default)


### Other updates
 - The `-d` flag (debug mode) now captures output from `rocotorun` in log files, such as job cards and job submission messages. This is not enabled by default because it is very slow.
 - Job monitoring can now be paused with `ctrl-c`, and a message on how to resume testing is printed to screen
 - All auto-generated experiment monitoring files now use the naming convention `WE2E_tests_DATETIME.yaml`
 - functions `check_task_get_extrn_ics()` and `check_task_get_extrn_lbcs()` are consolidated into a single `check_task_get_extrn_bcs()` function
 - If `TEST_VX_FCST_INPUT_BASEDIR` is not set for this platform, do not exit with an error, instead set to empty string
 - Added more detailed checks to ensure that an experiment is actually finished before script exits
 - Fixed broken capabilities (calculating "relative cost" and number of forecasts) in `WE2E_test_info.txt` file

### Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## TESTS CONDUCTED: 
I have tested on all platforms I have access to; I would appreciate help testing on other platforms.

- [x] hera.intel
  - Ran three identical suites (from `tests/WE2E/machine_suites/fundamental`) with the old `run_WE2E_tests.sh`, and with `run_WE2E_tests.py` in the default configuration, as well as using the ``--use_cron_to_relaunch`` option. All tests produced identical output.
  - Ran all test suite options (`fundamental`,`comprehensive`, and `all`), as well as individual tests and input files with test names to ensure expected behavior for all `-t` options
  - Quit script using `ctrl-c` at various points during the execution to ensure expected behavior
  - Ran `WE2E_summary.py` on several previous experiment directories to ensure expected output
- [x] orion.intel
  - Ran fundamental tests
- [x] cheyenne.intel
  - Ran fundamental tests
- [x] cheyenne.gnu
  - Ran fundamental tests
- [ ] gaea.intel
- [x] jet.intel
  - Ran all tests
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins

## DOCUMENTATION:
Documentation is updated for relevant WE2E capabilities that have been changed/added. A lot of documentation related to individual tests and suites is still outdated, but this will be updated at a future date with changes related to #587. 

This version of the docs can be viewed here; changes were made to chapters 10 and 12: https://ufs-srweather-app-mkavulich.readthedocs.io/en/latest/index.html

## ISSUE: 
 - Resolves #586 
 - Resolves #462
   - Users on Cheyenne can submit `./run_WE2E_tests.py` to avoid use of crontab; to run faster it can be run on a compute node with the `-p` option.
   - Users on Hera, where occasional failures occurred due to reaching a job submission limit, will no longer see this error (unless legacy crontab option is used)
 - Fixes #571 
 - Fixes #568 

## CHECKLIST
<!-- Add an X to check off a box. -->
- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] New and existing tests pass with my changes

## CONTRIBUTORS: 
@christinaholtNOAA Contributed a better way of setting relative dates (as used in test `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS`)

mkavulich committed:
Fix error introduced to pythonized WE2E script

mkavulich committed:
Fix failing "specify_template_filenames" test

mkavulich committed:
Add logic and instructions for neatly interrupting and resuming the monitor script

mkavulich committed:
Consolidate check_task_get_extrn_ics and check_task_get_extrn_lbcs into a single function: check_task_get_extrn_bcs

mkavulich committed:
Remove some unnecessary debug prints

mkavulich committed:
Initial version of job summary script

mkavulich committed:
Update experiment yamls to track task cores and walltime, update job_monitor to report core hours

mkavulich committed:
Print summary once all jobs are finished monitoring, add failsafe core hour calculation for missing NNODES variables, clean up formatting

mkavulich committed:
Move some functions into a new utils.py file to avoid circular dependencies

mkavulich committed:
Send rocotorun output to logging.debug, append logs rather than overwriting

mkavulich committed:
Add ability to update experiments in parallel

mkavulich committed:
Remove duplicate routine, add totals row to job summary

mkavulich committed:
More improvements

 - Remove another duplicate routine
 - Rename "job_summary" to "WE2E_summary"
 - Rename various auto-created yaml files to WE2E_tests_{TIME}.yaml for
consistency
 - Write "summary_{TIME}.txt" containing full report on WE2E tests

mkavulich committed:
Add final missing options:

 --opsroot, allows user to set NCO OPSROOT variable
 --print_test_details, allows user to print a pipe-delimited text file
(test_details.txt) analogous to previous WE2E_test_info.csv file

mkavulich committed:
- Remove incorrectly left-in exit call
 - Remove incorrectly left-in commented code
 - Add documentation for get_or_print_blank()
 - Some suggested changes from pylint

mkavulich committed:
New way of specifying relative time tests without system calls to `date`, courtesy of Christina Holt

mkavulich committed:
Some needed changes to address problems with parallel mode

 - Pass "refresh" flag correctly in parallel mode: only for first pass
through of tests list
 - Move second "rocotorun" call to immediately after for better chance
of creating rocoto db file prior to attempting to read
 - Only mark experiment in "error" state if it was not created after the
second pass through
 - Print warning message for the case where jobs are continuously not
submitted, giving users info in case they mis-configured their
experiment

mkavulich committed:
Add a final check using rocotostat to ensure that there are no un-submitted tasks remaining

mkavulich committed:
Fix logic for jobs in DEAD or UNKNOWN status, add logic for "FAILED" jobs, fix incorrect variable in error message

mkavulich committed:
Remove UNKNOWN from the status check; according to rocoto source code, UNKNOWN jobs will be retried, so we don't want to mark these as ERROR

mkavulich committed:
Some more fixes from final testing

 - Set VX_FCST_INPUT_BASEDIR to null string if not set on platform
   (rather than failing)
 - Print message about updating experiment on first go around; this
   allows user to see progress with parallel option
 - Only call compare_rocotostat() if job may be finished or stuck

mkavulich committed:
Start updating documentation

mkavulich committed:
Finish implementation of "print_test_details()"
 - Add stand-alone script for calling function without running full
test suite
 - Fix "calculate_cost" function and use it to calculate relative
test cost as in previous implementation
 - Add entry for number of forecasts in output

mkavulich committed:
For test details, rename everything to be more consistent with old
filenames. Also, fix link detection problem, and use default filename
when called from run script

mkavulich committed:
Continue updating documentation through first few sections of WE2E chapter

mkavulich committed:
Some more documentation updates

mkavulich committed:
Update gitignore for new/updated filenames

mkavulich committed:
Documentation for WE2E_summary.py script

mkavulich committed:
Revert behavior of rocotorun to only capture output if debug=True. This
will greatly increase the speed of runs on most platforms, but will have
the downside of not capturing all rocotorun output in log files,
including job cards and other messages.

mkavulich committed:
Fixes suggested by pylint

mkavulich committed:
The big moment: ditching the old shell version.

mkavulich committed:
Updates to Jenkins test script for new python workflow

mkavulich committed:
Fix unit test for new behavior of calculate_cost.py

mkavulich committed:
Fix unit test for real this time?

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,597 @@
+                core_hours = expt_dict[expt][task]['cores'] * expt_dict[expt][task]['walltime'] / 3600
+                expt_dict[expt][task]['exact_count'] = False
+            expt_dict[expt][task]['core_hours'] = round(core_hours,2)
+    return expt_dict
```
No need to return expt_dict here. It's updated in-place.

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,597 @@
+        #Update the experiment dictionary
+        logging.debug(f"Reading status of experiment {item}")
+        update_expt_status(expt_dict[item],item,True,False)
+    summary_file = f'WE2E_tests_{datetime.now().strftime("%Y%m%d%H%M%S")}.yaml'
```
This is bad indentation, I think.

christinaholtNOAA replied:
Oops. This is fine.

christinaholtNOAA commented on tests/WE2E/WE2E_summary.py:
```diff
@@ -0,0 +1,73 @@
+        raise ValueError(f'Bad arguments; run {__file__} -h for more information')
+
+    # Calculate core hours and update yaml
+    expt_dict = calculate_core_hours(expt_dict)
```
This should probably just update the dict in-place, no?

mkavulich committed:
Don't call rocotorun for WE2E_summary

mkavulich committed:
Add directory name to test summary

mkavulich committed:
- More general cleanup, including suggestions from pylint
 - Rename monitor_jobs.yaml to WE2E_tests.yaml to conform with naming
convention of auto-generated yaml files
 - Rename instances of "expt_dir" to "expts_dir" if it is a dictionary of
multiple experiment dictionaries; retain "expt_dir" for
single-experiment dictionaries
 - Remove duplicate copy of calculate_core_hours() that somehow pylint
doesnt catch??

mkavulich committed:
Fix missed "expt_dict" rename, widen test name column in summary file

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -283,121 +303,81 @@ def check_test(test: str) -> str:
-    Returns:
-        cfg_lbcs : Updated dictionary for task_get_extrn_lbcs section of test config
-    """
+    I_OR_L = ics_or_lbcs.upper()
```
`I_OR_L` is not a great variable name.

mkavulich committed:
Add missing column header for test info file

mkavulich commented:
## Notes on platform-specific usage
I wasn't sure the best place to put these notes so I'll paste them here for future reference.

While it is still possible to run WE2E tests using the crontab option, in general the new script is faster. To speed up the test submission even further, you can use the `-p=PROCS` option to run PROCS parallel processes.

### Cheyenne

Usage on Cheyenne was a big part of why this script was introduced (#462). Use of crontabs is discouraged by sysadmins on that platform, and there are fairly strict time and memory limits on the use of login nodes. While smaller test suites may run successfully using the login node, it is recommended to run tests on a login node.

```
cd ufs-srweather-app/tests/WE2E
qinteractive -l walltime=7200 -A account_name
module use ../../modulefiles/
module load wflow_cheyenne
conda activate regional_workflow
./run_WE2E_tests.py -m=cheyenne -a=account_name --tests=fundamental --expt_basedir=fundamental_parallel -p=32
```

The `qinteractive` command puts the user on a login node for an hour by default. The `-l walltime=7200` option extends this to 2 hours (7200 seconds). Then we can run `./run_WE2E_tests.py` with the `-p=32` options to make sure we utilize all 32 cores for job monitoring/submission.

### Hera
The rocotorun tasks run quite fast on Hera, so for most testing even non-parallel runs will submit jobs with little delay. For very large test suites, such as the `comprehensive` suite, running in parallel can give a significant speed-up, but due to limitations of the Hera system, it is not recommended to use more than 4 processes. Using more may result in rocotorun tasks being unable to initialize, which can result in errors in job submission.

MichaelLueken commented on .cicd/scripts/srw_test.sh:
```diff
@@ -38,58 +38,18 @@ else
+    opsroot=${nco_dir} | tee ${progress_file}
 
 # Progress file
 progress_file="${workspace}/we2e_test_results-${platform}-${SRW_COMPILER}.txt"
```
These lines should be removed, since they have been moved to lines 41 and 42.

MichaelLueken commented on .cicd/scripts/srw_test.sh:
```diff
@@ -38,58 +38,18 @@ else
+# Progress file
+progress_file="${workspace}/we2e_test_results-${platform}-${SRW_COMPILER}.txt"
 ./setup_WE2E_tests.sh ${platform} ${SRW_PROJECT} ${SRW_COMPILER} ${test_type} \
     expt_basedir=${we2e_experiment_base_dir} \
```
Both `expt_basedir` and `opsroot` should have `--` added to them:

--expt_basedir=${we2e_experiment_base_dir}
--opsroot=${nco_dir}

MichaelLueken commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -54,16 +54,16 @@ test_type=${4:-fundamental}
 opts=
 if [[ "$*" != *"debug"* ]]; then
-   opts="${opts} debug=TRUE"
+   opts="${opts} --debug=TRUE"
```
Since there is no boolean associated with `--debug` in `run_WE2E_tests.py`, `=TRUE` should be removed:

`opts="${opts} --debug`

MichaelLueken commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -54,16 +54,16 @@ test_type=${4:-fundamental}
 fi
 if [[ "$*" != *"verbose"* ]]; then
-   opts="${opts} verbose=TRUE"
+   opts="${opts} --verbose=TRUE"
```
Since there is no boolean associated with `--verbose` in `run_WE2E_tests.py`, `=TRUE` should be removed:

`opts="${opts} --verbose`

MichaelLueken reviewed (commented):
@mkavulich I was able to clone your branch and run the fundamental tests using `tests/WE2E/run_WE2E_tests.py` without issue.

I encountered issues while attempting to manually run the automated Jenkins tests (`.cicd/scripts/srw_test.sh`).  I have noted in the review the changes that I had to make in order to start the tests.  Following the required changes, the tests fail with the following:

```
calling function that monitors jobs, prints summary
Writing information for all experiments to WE2E_tests_20230307120101.yaml
Checking tests available for monitoring...
Starting experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta running
Updating database for experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta
03/07/23 12:01:02 CST :: FV3LAM_wflow.xml :: Workflow Manager Initialization failed.
03/07/23 12:01:02 CST :: FV3LAM_wflow.xml :: Permission denied - /.rocoto
/apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `initialize': No such file or directory - /.rocoto/1.3.3/FV3LAM_wflow/log.lock (Errno::ENOENT)
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `open'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `log'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/workflowengine.rb:78:in `rescue in initialize'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/workflowengine.rb:37:in `initialize'
	from /apps/contrib/rocoto/1.3.3/bin/../sbin/rocotorun.rb:34:in `new'
	from /apps/contrib/rocoto/1.3.3/bin/../sbin/rocotorun.rb:34:in `<main>'

03/07/23 12:01:02 CST :: FV3LAM_wflow.xml :: Workflow Manager Initialization failed.
03/07/23 12:01:02 CST :: FV3LAM_wflow.xml :: Permission denied - /.rocoto
/apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `initialize': No such file or directory - /.rocoto/1.3.3/FV3LAM_wflow/log.lock (Errno::ENOENT)
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `open'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/utilities.rb:187:in `log'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/workflowengine.rb:78:in `rescue in initialize'
	from /apps/contrib/rocoto/1.3.3/lib/workflowmgr/workflowengine.rb:37:in `initialize'
	from /apps/contrib/rocoto/1.3.3/bin/../sbin/rocotorun.rb:34:in `new'
	from /apps/contrib/rocoto/1.3.3/bin/../sbin/rocotorun.rb:34:in `<main>'

Reading database for experiment grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta, updating experiment dictionary

*********************************************************************
FATAL ERROR:
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: log.run_WE2E_tests
*********************************************************************

Traceback (most recent call last):
  File "/work/noaa/epic-ps/mlueken/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 534, in <module>
    run_we2e_tests(homedir,args)
  File "/work/noaa/epic-ps/mlueken/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 222, in run_we2e_tests
    monitor_file = monitor_jobs(monitor_yaml, monitor_file=monitor_file, procs=args.procs,
  File "/work/noaa/epic-ps/mlueken/ufs-srweather-app/tests/WE2E/monitor_jobs.py", line 50, in monitor_jobs
    expts_dict[expt] = update_expt_status(expts_dict[expt], expt, True, debug)
  File "/work/noaa/epic-ps/mlueken/ufs-srweather-app/tests/WE2E/utils.py", line 291, in update_expt_status
    for task in db:
UnboundLocalError: local variable 'db' referenced before assignment
```
Looking in `WE2E_tests_20230307120101.yaml`, I see:

```
### WARNING ###
### THIS FILE IS AUTO_GENERATED AND REGULARLY OVER-WRITTEN BY WORKFLOW SCRIPTS
### EDITS MAY RESULT IN MISBEHAVIOR OF EXPERIMENTS RUNNING
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta:
  expt_dir: /work/noaa/epic-ps/mlueken/ufs-srweather-app/expt_dirs/grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_RRFS_v1beta
  status: CREATED
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta:
  expt_dir: /work/noaa/epic-ps/mlueken/ufs-srweather-app/expt_dirs/grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
  status: CREATED
community_ensemble_008mems:
  expt_dir: /work/noaa/epic-ps/mlueken/ufs-srweather-app/expt_dirs/community_ensemble_008mems
  status: CREATED
deactivate_tasks:
  expt_dir: /work/noaa/epic-ps/mlueken/ufs-srweather-app/expt_dirs/deactivate_tasks
  status: CREATED
inline_post:
  expt_dir: /work/noaa/epic-ps/mlueken/ufs-srweather-app/expt_dirs/inline_post
  status: CREATED
```
It seems as though the step to start the tests running didn't work correctly, which led the `FV3LAM_wflow.db` file not to be generated.  Do you have any idea why this might be occurring for the `.cicd/scripts/srw_test.sh` script?

mkavulich commented:
@MichaelLueken Based on what I'm seeing in the error messages you posted, it looks like there is a missing or uninitialized variable specifying where the .rocoto directory is. On Hera and Cheyenne this is the $HOME directory, maybe $HOME is not set on the Jenkins machine?

mkavulich committed:
Fixes to Jenkins testing scripts from Mike Lueken

mkavulich committed:
If database is not loaded, need to return every time, even if a tolerable error.

MichaelLueken commented:
@mkavulich Thanks for making the changes to allow `.cicd/scripts/srw_test.sh` to run the updated `run_WE2E_tests.py` script.  I'll see what I can find with respect to how rocoto is set for Jenkins.  As I understand it currently, there are Jenkins runners deployed on all Tier-1 machines, linked to the EPIC role accounts on the various machines.  I don't think that there is currently a variable for specifying where the .rocoto directory is.

MichaelLueken commented:
@mkavulich If you would like to try and replicate the errors by manually running the Jenkins scripts, you will need to set the following parameters (this example is for Orion, but are still used for other machines as well):

```
export SRW_PLATFORM="orion"
export SRW_COMPILER="intel"
export WORKSPACE="/work/noaa/epic-ps/mlueken/ufs-srweather-app"
export SRW_WE2E_COMPREHENSIVE_TESTS="False"
export SRW_PROJECT="epic-ps"
```

Where `SRW_PLATFORM` is the machine you are running on, `SRW_COMPILER` is the compiler, `WORKSPACE` is the path to your local copy of ufs-srweather-app,  `SRW_WE2E_COMPREHENSIVE_TESTS` is a flag used to turn on or off the comprehensive tests, and `SRW_PROJECT` is the account.

Once these environmental variables are set, then you will need to rebuild the SRW App using `./srw_build.sh`.  Once built, you can submit the tests using `./srw_test.sh`.  Both of these commands will need to be used in the `.cicd/scripts` directory.

mkavulich commented:
@MichaelLueken Thanks for the instructions, I was just about to ask! I will let you know what I find.

mkavulich commented:
@MichaelLueken I have tracked down the problem to [this line in the setup script](https://github.com/mkavulich/ufs-srweather-app/blob/feature/replace_sh_WE2E_script/tests/WE2E/setup_WE2E_tests.sh#L2) that seems to un-set the $HOME directory. It came from [a bug fix PR #459 by Daniel Abdi](https://github.com/ufs-community/ufs-srweather-app/pull/459/files#diff-7153880786409fadf07acb91061319b86e7c7a30b9d1f3bb16e2ea3222859a49), but I do not understand exactly what this line is doing.

@danielabdi-noaa Do you remember what this line does and/or why it was included? Removing this line seems to fix the problem we're seeing.

mkavulich commented:
@MichaelLueken I confirmed that removing that line results in a successful Jenkins test on Jet. We'll need to wait to hear back from @danielabdi-noaa as to what the purpose of that line is; we need to make sure at the very least that the `$HOME` variable is set in the environment when running the script; this appears to be a limitation of rocoto, not our workflow.

danielabdi-noaa commented:
@mkavulich If I recall correctly, this line is added so that the `setup_WE2E_tests.sh` script gets a clean environment smiliar to one you would get with slurm's `--export=NONE`. Without it the setup script failed when you execute it from a shell where the conda environment is activated i.e.
```
(regional_workflow) $ ./setup_WE2_tests.sh
```
vs
```
$ ./setup_WE2_tests.sh
```
The first one was not working because you have conda activated twice, within and outside the setup script and paths were mixed up and it was not able to find `jinja2` or `yaml` IIRC. I can not find the chat we had with @MichaelLueken but there was an issue then for sure. Things may have changed since then with upgrades to modulefiles so you can try to remove it if you can't find other solution. Personally I like it since it will make sure the script behaves the same whether conda is activated outside the script or not.

mkavulich committed:
Rocoto requires /home/Michael.Kavulich to be set to a writable path in the user's environment, so we have to pass it as an argument to setup_WE2E_tests.sh to be exported prior to running the WE2E test script

mkavulich commented:
@danielabdi-noaa Thanks for the info! With that in mind, I was able to implement a fix that still keeps a clean environment, just passing $HOME as an argument to be exported later. This has allowed the Jenkins script to run successfully on Jet!

@MichaelLueken You can try the script again now and it should complete successfully. Just note that if you run on Hera you may see random failures; the reason for this is due to an NCO mode problem described in an issue I just opened (#652)

MichaelLueken commented:
@mkavulich and @danielabdi-noaa Thank you both very much for identifying the cause of the issue and correcting it!  The manual run of the Jenkins test script I ran on Orion successfully runs now.

MichaelLueken reviewed (approved):
@mkavulich Manually running `run_WE2E_tests.py`, manually running the Jenkins test script - `.cicd/scripts/srw_test.sh`, and relaunching tests using `monitor_jobs.py` were all successful.  Approving this work now.

MichaelLueken commented:
Please note that the Jenkins failure is due to issue #635.  The Jenkins test is failing for Orion.  I have manually submitted the tests on Orion and they have successfully passed, as noted above.

MichaelLueken commented:
@mkavulich Additional issues have been encountered with the Jenkins tests.

During the tarring of the log files at the end of the tests, the following is being encountered:

`*/log.launch_FV3LAM_wflow: Cannot stat: No such file or directory`

This is causing the testing phase to fail, since this file is required in the tarring step.

The Jenkinsfile includes the following line:
https://github.com/ufs-community/ufs-srweather-app/blob/ff6f103b451bb0886b90afdc4a2a7a80f877c36d/.cicd/Jenkinsfile#L180

Since `log.launch_FV3LAM_wflow` is no longer part of the workflow, this entry on line 180 of the .cicd/Jenkinsfile will need to be removed.

mkavulich committed:
Update archiving of relevant log files in Jenkinsfile

mkavulich commented:
@MichaelLueken thanks for the info, I removed the now-non-existent log file from the archive, and also added new log and status files to the archive. Let me know if this still does not work.

danielabdi-noaa commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -1,5 +1,10 @@
+# `exec -c` runs this script with clean environment; this avoids some problems
+# with double-loading conda environments. Since we do need $HOME to be set for
+# rocoto to run properly, pass it as an argument and export it later
+
```
Ok good fix!

danielabdi-noaa commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -26,8 +31,9 @@
-  echo "Usage: $0 machine account [compiler] [test_type] [others] | -h"
+  echo "Usage: $0 homedir machine account [compiler] [test_type] [others] | -h"
   echo
+  echo "       homedir       [required] user's home directory; this space must be writable"
```
I don't think the user needs to know about the implictly passed $HOME directory, so probably remove this and the one above on the `Usage: $0 homedir` ?

danielabdi-noaa commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -42,43 +48,54 @@ function usage {
+echo $machine
+echo $account
+echo $compiler
+echo $test_type
```
These are probably debugging prints?

danielabdi-noaa commented on .cicd/scripts/srw_test.sh:
```diff
@@ -38,58 +38,15 @@ else
+./setup_WE2E_tests.sh ${platform} ${SRW_PROJECT} ${SRW_COMPILER} ${test_type} \
+    --expt_basedir=${we2e_experiment_base_dir} \
+    --opsroot=${nco_dir} | tee ${progress_file}
 
```
I don't know how the new `run_WE2E` works but just wanted to make sure the various delays are taken care of
- initial delay of 300 sec 
- polling frequency of 60 sec
- after completion delay of 600 sec. This one is to make sure jenkins does not delete experiment directory prematurely.

mkavulich replied:
Unless I'm missing something, none of those delays should be relevant with the new script. 

1. The `./run_WE2E_tests.py` script takes care of both the setting up and monitoring of tests, so no initial delay is needed
2. The new script serially cycles through active experiments with no delay
3. The new script will not finish until all tasks have been confirmed complete. 

One thing that may need to change is that for larger tests (such as, if the comprehensive tests are re-enabled), it may be worth using the new `-p` parallel option. Because the Jenkins tests call the script with `-d`, calls to rocotorun are fairly slow, taking ~30 seconds each (and there are two calls to rocotorun per test check). But this should be done cautiously depending on the platform: too many parallel tasks may overload the login node if that's where this script is running.

danielabdi-noaa replied:
For the third one, note that the existing `./run_WE2E_Tests.sh` also does not return until all tasks are complete, so if it was needed for the shell script it may also be needed for the python script. I vaguely recall that server side jenkins code by @jessemcfarland  deletes experiment directories as soon as `srw_tests.sh` finishes. This could happen before all cronjobs are removed, since they need to be called one last time for them to be removed from the crontable.

The first one, initial delay, is required to given enough time for atleast one workflow to be launched. If `run_WE2E_tests.py` does not exit thinking all jobs completed, when in reality no jobs were launched after say 4 mins, then it should be fine.

The second one is ok i think since it looks like the behaviour of checking experiment status changed. The old `get_expt_status.sh` was time consuming so it checks status of experiments every minute and print to the screen.

mkavulich replied:
>note that the existing ./run_WE2E_Tests.sh also does not return until all tasks are complete

The `launch_FV3LAM_wflow.sh` script *tried* to implement this but it was done in quite a hacky way, so there were instances where the script would assume the experiment was done when there were still jobs running (for example, if there was a failed task but other tasks were still running). The new implementation does not use that script (does not use cron jobs either) and so does not suffer from that problem. It will only exit if it has verified that all jobs are done, via reading the rocoto database file directly.

>If run_WE2E_tests.py does not exit thinking all jobs completed, when in reality no jobs were launched after say 4 mins, then it should be fine.

This is correct, the `run_WE2E_tests.py` both submits and monitors jobs, so unless something goes horribly wrong there is no possibility of "desync" there.

I won't promise that I have anticipated *all* potential problems, but if any of these problems you mention (or similar ones) happen in the future, I can fix it in the `run_WE2E_tests.py` script itself, we shouldn't rely on hand-wavy delays to avoid problems anyway.

danielabdi-noaa replied:
Ok I did not realize cron is no longer being used. I agree the initial/final delays are not required in this case.

danielabdi-noaa commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -1,5 +1,10 @@
 #!/usr/bin/env bash
```
Could you update the usage comments of this script below? I was able to run fundamental tests witch
```
./setup_WE2E_tests.sh hera zrtrr intel fundamental
```
But not a custom test file `custom.hera.intel.com`
```
./setup_WE2E_tests.sh hera zrtrr intel custom
```

mkavulich replied:
The script should be able to work with any custom file name, suite name, or test name; I confirmed this worked through the `./setup_WE2E_tests.sh` script with my own test on Jet. Can you send me the command that is failing for you and the directory? The above two commands appear to be the same.

danielabdi-noaa replied:
@mkavulich I think test files like `custom.hera.intel.com` are not recognized because such patterns are only parsed for `fundamental` and `comprehensive` tests according to the logic in line 72 of `run_WE2E_tests.py`
```
$ ./setup_WE2E_tests.sh hera zrtrr intel custom
/home/Daniel.Abdi
hera
zrtrr
intel
custom

Modules based on Lua: Version 8.5.2  2021-05-12 12:44 -05:00
    by Robert McLay mclay@tacc.utexas.edu


Currently Loaded Modules:
  1) rocoto/1.3.3   2) miniconda3/4.12.0   3) wflow_hera

 

Logging set up successfully
Checking if ['custom'] is a valid test name
Checking if ['custom'] is a valid test suite
Checking if ['custom'] is a file containing test names

*********************************************************************
FATAL ERROR:
Experiment generation failed. See the error message(s) printed below.
For more detailed information, check the log file from the workflow
generation script: log.run_WE2E_tests
*********************************************************************

Traceback (most recent call last):
  File "/scratch2/BMC/gsd-hpcs/Daniel.Abdi/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 534, in <module>
    run_we2e_tests(homedir,args)
  File "/scratch2/BMC/gsd-hpcs/Daniel.Abdi/ufs-srweather-app/tests/WE2E/./run_WE2E_tests.py", line 109, in run_we2e_tests
    raise FileNotFoundError(dedent(f"""
FileNotFoundError: 
The specified 'tests' argument '['custom']'
does not appear to be a valid test name, a valid test suite, or a file
containing valid test names.

Check your inputs and try again.
```

mkavulich replied:
@danielabdi-noaa Sorry, I still don't understand what behavior you are expecting. It appears you gave the script `custom` as an argument; if that's not a file name then I'm not aware of what functionality that's supposed to have.

danielabdi-noaa replied:
The file I have is `custom.hera.intel.com` under `machine_suites` folder not `custom`. With the new `setup_WE2E_tests.sh`, none of the below two calls work
```
./setup_WE2E_tests.sh hera zrtrr intel custom.hera.intel.com
./setup_WE2E_tests.sh hera zrtrr intel custom
```
while with the old one, they both do. Line 72 of `run_WE2E_tests.py` is a new addition that was absent in the old one, and I believe is the cause of this problem. I can understand if the behaviour is changed now, but that should be documented if so.

mkavulich replied:
Thanks for the explanation, sorry it took me a while to understand. The way you are using the old script is not documented (at least, not anywhere I'm aware of) so I did not account for it when consolidating the three test arguments (`test_type`, `tests_file`, `test_name`) into a single argument (`tests`). I would argue that your use is unintended behavior and should not be carried forward to the new script. 

The documented/intended way to specify a custom test list file is to provide the full or relative path location of the file name you wish to read. So to run a file named `custom.hera.intel.com` under `machine_suites`, you would run

`./setup_WE2E_tests.sh hera zrtrr intel machine_suites/custom.hera.intel.com`

dshawul replied:
@mkavulich Ok sounds good. In that case the usage of the script should be modified accordingly
```
echo "       test_type     [optional] test type: fundamental or comprehensive or all or any other name"
```
The `test_type` can not be "any other name" now, and also due to the consoldation of the three arguments maybe add also that you can pass a specific test case name or a path to a file containing list of tests.

mkavulich replied:
Thanks Daniel, I have updated the usage statement in `./setup_WE2E_tests.sh` to accurately reflect the new argument behavior

mkavulich committed:
Fixes suggested by Daniel

mkavulich committed:
Fix usage instructions for more flexible "tests" argument

christinaholtNOAA commented on tests/WE2E/WE2E_summary.py:
```diff
@@ -0,0 +1,71 @@
+
+
+
+if __name__ == "__main__":
```
It's been helpful for testing other scripts if all the stuff in the "main" part of the script goes into a function, and the only thing the script does is call that function. For example:

```
def create_summary(args):
     # All the stuff below this line


if __name__ == "__main__":
    create_summary(sys.argv[1:])
```

christinaholtNOAA replied:
This one doesn't seem to have been addressed.

mkavulich replied:
I re-pushed the change here: https://github.com/ufs-community/ufs-srweather-app/pull/637/commits/8f5525d152965511d65b17a575d973ac9d1d6626, it was un-intentionally reverted.

christinaholtNOAA replied:
After a quick in person chat, I think I'm caught up. This suggestion was an organization suggestion for ease of future testing, and is out of scope for this PR given that many other scripts with user interfaces need similar care.

christinaholtNOAA commented on tests/WE2E/monitor_jobs.py:
```diff
@@ -3,217 +3,100 @@
 
 from check_python_version import check_python_version
 
+from WE2E_summary import print_WE2E_summary
```
`print_WE2E_summary` seemed to be imported from `utils` in this script above.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -162,32 +173,32 @@ def run_we2e_tests(homedir, args) -> None:
+                       "based on specified command-line arguments:\n")
         logging.debug(cfg_to_yaml_str(test_cfg))
-        with open(ushdir + "/config.yaml","w") as f:
+        with open(ushdir + "/config.yaml","w", encoding="utf-8") as f:
```
`os.path.join()` is more portable, in general.

christinaholtNOAA commented on tests/WE2E/run_WE2E_tests.py:
```diff
@@ -283,121 +303,81 @@ def check_test(test: str) -> str:
+                    dflt[f'task_get_extrn_{ics_or_lbcs}'][f'FV3GFS_FILE_FMT_{I_OR_L}']
+        cfg_bcs[f'EXTRN_MDL_SOURCE_BASEDIR_{I_OR_L}'] = \
+                f"{mach['platform']['TEST_EXTRN_MDL_SOURCE_BASEDIR']}/"\
+                f"{cfg_bcs[f'EXTRN_MDL_NAME_{I_OR_L}']}/{cfg_bcs[f'FV3GFS_FILE_FMT_{I_OR_L}']}/${{yyyymmddhh}}"
```
There's a chance the `os.path.join()` route here would make this much more readable...and portable.

christinaholtNOAA commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -42,43 +48,46 @@ function usage {
-account=$2
-compiler=${3:-intel}
-test_type=${4:-fundamental}
+homedir=$1
```
This shift in order of the input command line args doesn't seem to be reflected in the usage statement.

mkavulich replied:
This was changed per @danielabdi-noaa's comments; since this script is re-called automatically with homedir and it is not a user-supplied argument, the usage statement shouldn't include it

christinaholtNOAA commented on tests/WE2E/setup_WE2E_tests.sh:
```diff
@@ -26,14 +31,15 @@
+  echo "       compiler  [optional] compiler used to build binaries (intel or gnu)"
+  echo "       tests     [optional] tests to run: can be a suite (all|comprehensive|fundamental)
+                            a filename, or a test name"
+  echo "       others    [optional] All other arguments are forwarded to run_WE2E_tests.sh"
```
This one should be .py now?

christinaholtNOAA commented on tests/WE2E/test_configs/wflow_features/config.specify_template_filenames.yaml:
```diff
@@ -5,11 +5,11 @@ metadata:
-  MODEL_CONFIG_TMPL_FN: model_configure
-  NEMS_CONFIG_TMPL_FN: nems.configure
+  MODEL_CONFIG_FN: model_configure
+  NEMS_CONFIG_FN: nems.configure
```
I'm curious if this test can just be deleted....it no longer serves the purpose since these values are the default values

mkavulich replied:
This test will be removed in my next PR (related to Issue #587); a few test updates are needed like this, and I'd rather keep them compartmentalized for easier review. These changes were simply to let this test actually pass if run.

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+                continue
+            status = expts_dict[expt][task]["status"]
+            walltime = expts_dict[expt][task]["walltime"]
+            expt_details.append(f'{task[:40]:<40s}  {status:<12s} {walltime:>10.1f}')
```
Lots of 40s as magic numbers here if we every want to make that wider...

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+                             expts_dict[expt][task]['walltime'] / 3600
+                expts_dict[expt][task]['exact_count'] = False
+            expts_dict[expt][task]['core_hours'] = round(core_hours,2)
+    return expts_dict
```
You're update `expts_dict` in place. No need to return it.

mkavulich replied:
Trying to implement this resulted in breaking the parallel (starmap) calls to this function. I'd rather leave this how it is since it's technically correct.

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+                           workflow by calling rocotorun. If simply generating a report, set this
+                           to False
+    Returns:
+        dict: The updated experiment dictionary.
```
No need to return the dict. It's being updated in place. (this applies to others below, too, but I'll quite harping here)

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+        expts_dict (dict): A dictionary containing information for all experiments
+        procs      (int): The number of parallel processes
+        refresh   (bool): "Refresh" flag to pass to update_expt_status()
+        debug     (bool): Will capture all output from rocotorun. This will allow information such
```
Silly formatting pet peeve...either keep the colons aligned or use one space between variable name and the data type.

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+    links = dict()
+    for testfile in testfiles:
+        # Calculate relative cost of test based on config settings using legacy script
+        cost_array = calculate_cost(testfile)
```
Is this array an array or a list? Maybe even a list of lists?

There are a bunch of numbers referencing list items in this section. Perhaps it's helpful to the user here to see a comment on how you expect the cost_array to be structured? Or even split them out into variables like this to improve readability:
```
test_cost, _, baseline_cost = cost_array[1:3]
cost = test_cost / baseline_cost
```

mkavulich replied:
This was a product of legacy code that I did not feel was in the scope of this PR to address. Ideally `calculate_cost` should be completely re-written (I already had to make significant changes to get it working again) but this would require a lot of effort that I feel is best placed elsewhere

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+    for key in links.keys():
+        link = links[key]
+        testdict[link[2]]["alternate_name"] = link[0]
+        testdict[link[2]]["alternate_directory_name"] = link[1]
```
This could be made a bit more readable:

```
for key, link in links.items():
    alt_name, alt_dir, link_name = link
    .....
```

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_run_fcst','DT_ATMOS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'task_get_extrn_lbcs',
+                                                'LBC_SPEC_INTVL_HRS'))
+            f.write(f"{d}" + get_or_print_blank(testdict[expt],'global','NUM_ENS_MEMBERS') + "\n")
```
This bit is pretty clunky. Just to avoid repetition and inefficient calls to a function that does what a built-in does, you might consider:

```
key_pairs = [
 ('workflow', 'PREDEF_GRID_NAME'),
 ('workflow', 'CCPP_PHYS_SUITE'),
  ...
  ]

for key1, key2 in key_pairs:
    f.write(f"{d} {testdict[expt].get(key1, {}).get(key2, '')}"
```

christinaholtNOAA commented on tests/WE2E/utils.py:
```diff
@@ -0,0 +1,584 @@
+    # Call rocotostat and store output
+    rocoto_db = f"{expt_dict['expt_dir']}/FV3LAM_wflow.db"
+    rocotorun_cmd = ["rocotostat", f"-w {expt_dict['expt_dir']}/FV3LAM_wflow.xml",
+                     f"-d {rocoto_db}", "-v 10"]
```
Why not pull the xml path out for similar readability?

christinaholtNOAA reviewed (commented):
Sorry for the delayed review. I left some comments/questions below.

gsketefian commented:
@mkavulich I looked through the changes and they seem fine.  I have a few questions that I could find the answers to by carefully looking at the code, but I figure just asking is much faster:
1) With the current version of `run_WE2E_tests.py` and accompanying scripts, I've noticed that while the script is re-submitting jobs, if I go into an experiment directory and call `rocotorun` or `launch_FV3LAM_wflow.sh`, I get the following error:
```
$ ./launch_FV3LAM_wflow.sh 

Modules based on Lua: Version 8.5.2  2021-05-12 12:44 -05:00
    by Robert McLay mclay@tacc.utexas.edu

Currently Loaded Modules:
  1) rocoto/1.3.3   2) miniconda3/4.12.0   3) wflow_hera

FATAL ERROR: 
ERROR:
  From script:  "launch_FV3LAM_wflow.sh"
  Full path to script:  "/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/DTC_ensemble_task_pythonize/ufs-srweather-app/ush/launch_FV3LAM_wflow.sh"
Call to "rocotorun" failed with return code 1.
Exiting with nonzero status.
```
This doesn't always happen, and it may depend on the task the experiment is currently on.  Have you encountered this, and do you know if the changes in this PR get around this?
2) Is the `launch_FV3LAM_wflow.sh` script being kept, and if not, is there a replacement for it that can be called from within an experiment directory (maybe via a symlink) that will relaunch the workflow?  This is handy for me when I'm debugging a single task and want to relaunch it over and over again and get its status.  
3) Is `WE2E_summary.py` the replacement for `get_expts_status.sh`?
Thanks.

mkavulich commented:
@gsketefian Thanks for your comments and questions.

1. This is a limitation of the `launch_FV3LAM_wflow.sh` script and how it interacts with rocoto. Rocoto locks the `.db` file while rocotorun is being run, so if you try to run two rocotorun processes simultaneously, the second will fail with the message `WARNING: Workflow is locked by pid xxxxxx on host xx.xxx.xx.xx` and exit code 1. So if the `./run_WE2E_tests.py` or `./monitor_jobs.py` scripts are using rocotorun while you run that script, it will fail. The same thing would happen if you tried to run `launch_FV3LAM_wflow` twice simultaneously. Is there a use case I'm missing that would require running `launch_FV3LAM_wflow` while the `monitor_jobs` function is actively monitoring/submitting jobs?
2. `launch_FV3LAM_wflow.sh` is not being removed with this PR (I was only addressing the WE2E tests in this PR, and I think this script has uses outside of that testing framework), but if you want to get the status of a job `WE2E_summary.py` will give you the summary of the currently running jobs (you can give it a yaml file or a top-level experiment directory as an argument). 
If you need the ability to see the status of an active experiment while you're in that directory, is there a reason that simply calling `rocotostat` doesn't give you want you need?
3. `WE2E_summary.py` is not a 1-to-1 replacement of `get_expts_status.sh`, but it should give the same relevant information. It does not not have the `num_log_lines` argument (because it reads rocoto `.db` files directly rather than relying on parsing a log file) and it does not have the `launch_wflows` argument (`WE2E_summary.py` does not advance the workflow with rocotorun, it only reads and reports the current status). 
If the ability to manually advance the workflow one step at a time is needed it can be added. But I was not anticipating that use being necessary.

gsketefian commented:
@mkavulich:
1. A use case is when you launch a bunch of tests with `run_WE2E_tests.py`, and say one of them starts to fail (some tasks report back as failed), and you want to stop resubmitting tasks for that experiment and go to its directory and run one task at a time to debug (while allowing the other experiment to continue).  Do `run_WE2E_tests.py` and `monitor_jobs.py` lock the `.db` files of all experiments while they're running?  Or do they read it, free it for a bit, then read it again, etc?  I have a feeling it's the latter since I don't always get that error.
2. As I step through the tasks, I like to get the most updated status of the experiments, so I like to run `launch_FV3LAM_wflow.sh` because it calls `rocotorun` before `rocotostat` and stores the results in a log file, which is helpful if I want to see the progress (e.g. what tasks were called from one re-launch to the next).  I realize these can all be done from the command line, but it's nice to have one script do it all.  As long as `launch_FV3LAM_wflow.sh` is not being removed, it should be fine.
3. Sounds like `WE2E_summary.py ` will be work for my purposes.

Currently, if I want to step one task at a time with an experiment is to first create it with `run_WE2E_tests.py`, then `Ctr-C` out of it once it's done creating the experiment directories, then go to the experiment directory and use `launch_FV3LAM_wflow.sh` to go one or two tasks at a time while I debug the xml, the ex-script, j-job, METplus conf file, etc.  I'll just keep using that approach.
Thanks!

mkavulich commented:
@gsketefian To answer your first question, the new scripts cycle through each experiment serially, running rocotorun on one experiment at a time, with a 5 second delay between loops. It is only while rocotorun is being called that the rocoto database is locked.

As a side note, if one experiment fails, then the monitor_jobs function will not continue running rocotorun in that directory. So if that is your use case you shouldn't see conflicts when using the launch_FV3LAM_wflow.sh script unless something weird is happening.

mkavulich committed:
Address PR comments

 - Correct import location for print_WE2E_summary
 - Use os.path.join() for path strings
 - Correct script name
 - Set global variables for column width in job summary

Also a bug fix for cases where variable definitions file doesn't exist
(can occur if experiment is moved or re-created after yaml generation)

mkavulich committed:
A couple bug fixes from latest changes

mkavulich committed:
Addressing more review comments: Update dicts in place

mkavulich committed:
Revert "Addressing more review comments: Update dicts in place"

This reverts commit b88c4a7dc569a4420a0fafc0918fe80be24a174c.

mkavulich committed:
Final set of review comments

mkavulich commented:
@christinaholtNOAA I believe I have addressed all your comments. Let me know if you have anything further.

mkavulich committed:
Un-revert intended change to flow of WE2E_summary.py

christinaholtNOAA reviewed (commented):
LGTM. Just one comment below.

mkavulich committed:
Revert "Un-revert intended change to flow of WE2E_summary.py"

This reverts commit 8f5525d152965511d65b17a575d973ac9d1d6626.

christinaholtNOAA reviewed (approved):
LGTM.

MichaelLueken commented:
@mkavulich All Jenkins tests successfully passed. The Orion tests ultimately sat in queue for several hours (over eleven hours), leading to a time-out for the Build and Test check.  Will now move forward with merging this PR.
